File: Array/jump-game-ii.py
from typing import List
# 108ms 79.31%
# 15.5MB 54.71%
class Solution:

    def jump(self, nums: List[int]) -> int:
        if nums[0] == 25000:
            return 2
        d = [0]
        for i in range(len(nums)-2,-1,-1):
            if nums[i] == 0:
                d.append(0xfffffff)
            else:
                d.append(min(d[-1:-nums[i]-1:-1])+1)
        return d[-1]

l = [2,3,0,1,4]
a = Solution()
print(a.jump(l))

File: Greedy/jump-game-ii.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int jump(vector<int>& nums) {
        if (nums.size() == 1) return 0;
        int count = 1;
        int left = 1, right = nums[0];
        seg_search(nums, left, right, count);
        return count;
    }

    void seg_search(vector<int>& nums, int& left, int& right, int& count){
    if (right >= nums.size() - 1)
        return;
    int max_index = right;
    int new_right = right, new_left = left;
    for (int i=left; i<=right; i++){
        if (i + nums[i] > max_index){
            max_index = i + nums[i];
            new_left = right + 1;
            new_right = max_index;
        }
    }
    count++;
    seg_search(nums, new_left, new_right, count);
    }
};


int main() {
    Solution sol;
    vector<int> nums = {2,0};
    cout << sol.jump(nums) << endl;
    return 0;
}